dwave/optimization/include/dwave-optimization/array.hpp

```diff
@@ -237,338 +237,473 @@ struct Update {
-                distance += mul * steps[dim];
-                mul *= shape[dim];
-            }
+            // Return the pointer offset (in bytes) relative to the current
+            // position in order to increment the iterator once.
+            difference_type increment1() {
+                difference_type offset = 0;
 
-            return distance;
-        }
+                ssize_t dim = ndim - 1;
+                for (; dim >= 1; --dim) {
+                    offset += strides[dim];
 
-        ssize_t ndim;
-        const ssize_t* shape;
-        const ssize_t* strides;
+                    // we don't need to advance any other dimensions
+                    if (++loc[dim - 1] < shape[dim]) break;
 
-        // loc is of length max(ndim - 1, 0) because we don't track our location in the
-        // 0th dimension
-        std::unique_ptr<ssize_t[]> loc;
-    };
+                    offset -= strides[dim] * shape[dim];  // zero out this dim
+                    loc[dim - 1] = 0;
+                }
+                if (dim == 0) {
+                    // For dim 0, we don't care if we're past the end of the shape.
+                    // This allows us to work for dynamic arrays as well.
+                    offset += strides[dim];
+                }
 
-    // unique_ptr is not trivial so the best we can have for ShapeInfo is nothrow
-    static_assert(std::is_nothrow_copy_constructible<ShapeInfo>::value);
-    static_assert(std::is_nothrow_move_constructible<ShapeInfo>::value);
-    static_assert(std::is_nothrow_copy_assignable<ShapeInfo>::value);
-    static_assert(std::is_nothrow_move_assignable<ShapeInfo>::value);
+                return offset;
+            }
 
-    // shape_ == nullptr => the array is contiguous. Otherwise the stride information
-    // will be encoded in the `shape_`.
-    std::unique_ptr<ShapeInfo> shape_ = nullptr;
-};
+            // Return the pointer offset (in bytes) relative to the current
+            // position in order to increment the iterator n times.
+            // n can be negative.
+            difference_type increment(ssize_t n) {
+                // handle a few simple cases with faster implementations
+                if (n == 0) return 0;
+                if (n == +1) return increment1();
+                if (n == -1) return decrement1();
+
+                difference_type offset = 0;  // the number of bytes we need to move
+
+                // working from right-to-left, figure out how many steps in each
+                // axis. We handle axis 0 as a special case
+                {
+                    // We'll be using std::div() over ssize_t, so we'll store our
+                    // current location in the struct returned by it.
+                    // Unfortunately std::div() is not templated, but overloaded,
+                    // so we use decltype instead.
+                    decltype(std::div(ssize_t(), ssize_t())) qr{ .quot = n, .rem = 0 };
```

Comment:
Given how annoying it is to use `std::div` for this, would it be cleaner to just use `/` and `%` and a temporary variable? Is it that you need consistency for a negative remainder? Seems like you're already accounting for that.